### DISCUSSION TITLE
Dynamic endpoint function signature

### DISCUSSION THREAD

**Discussion by marcovc** [fastapi/fastapi#7107] [Questions] [answered]
Hi,

Is it possible to create an endpoint from a dict of parameters, something like:

```python3

params = {"param1": int, "param2": str }

@app.get("/root")
def root(**params):
  pass
```

?

(the above doesn't work, the docs for example say that root takes one parameter called `params`)

Thank you!
Marco

**Comment by ycd** [comment]
What about using [Classes as Dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/)

Do you have any specific reasons that restrict you to use a dictionary?

**Comment by marcovc** [comment]
Thanks @ycd. I'm not sure if that solves it for me though. This allows me to do

```python3
class Args:
  def __init__(self. param1: int, param2: str):
    pass

@app.get("/root")
def root(args: Args = Depends()):
  pass
```

but what I would really like is to be able to build Args dynamically at runtime (adding param1 and param2).

**Comment by marcovc** [comment]
I was able to do this using a very ugly hack involving `exec()`. Closing.

**Comment by Mause** [comment]
I suggest you look at the makefun library, it works quite well for this use

https://pypi.org/project/makefun/

**Comment by ycd** [comment]
@marcovc can you share the snippet?

**Comment by marcovc** [comment, marked as answer]
@ycd sure, it is along these lines:

```python3
class Args:
  @classmethod
  def create(cls, params):
    params_str = ",".join(f"{k}: {v}" for k, v in params)
    exec(f"def fn(self, {params_str}): pass;\ncls.__init__ = fn")
    return cls

@app.get("/root")
def root(args: Args.create(run_time_params) = Depends()):
  pass
```

@Mause I'll have a look, thanks!

**Comment by marcovc** [comment]
Here's a solution with @Mause suggestion:

```python3
from makefun import with_signature

params_str = ",".join(f"{k}: {v}" for k, v in run_time_params)

class Args:
  @with_signature(f"__init__(self, {params_str}"):
  def fn(*args, **kwargs):
    pass

@app.get("/root")
def root(args: Args = Depends()):
  pass
```

Thanks both!

**Comment by tiangolo** [comment]
Thanks for the help here @Mause and @ycd ! :clap: :bow: 

Thanks for reporting back and closing the issue @marcovc  :+1:
